### PR TITLE
Feature refactor tests continue

### PR DESCRIPTION
**Pull request by mrfishball**
refactor specs

### CODE REVIEW

**Review by cmvandrevala — APPROVED** [review]
Just a heads up, I wouldn't call this a "Feature" in the PR description - you are just refactoring code in this PR.

File: Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift [hunk 1]
```diff
@@ -5,48 +5,104 @@ import Foundation
 
 class HttpResponseSpec: QuickSpec {
     override func spec() {
-        describe("HttpResponse") {
-            var responseBuilder = ResponseBuilder()
+        describe("A Reponse object") {
```
Comment (ericmeyer) [inline Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift:8]: Much better specs overall.

File: Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift [hunk 2]
```diff
@@ -5,48 +5,104 @@ import Foundation
...
-                    .with(statusCode: .ok)
-                    .with(contentType: .text)
-                    .with(body: String())
-                    .build()
-                
-                expect(responseOne).to(equal(responseTwo))
+            context("when compares with another response object") {
+                it("is true when attributes match") {
+                    var responseBuilder = summonBuilder()
+
```
Comment (ericmeyer) [inline Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift]: I personally would not add a newline here, but that's just my style preference. To me, the line above the newline and the section below the newline are both part of the setup of the individual test.

File: Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift [hunk 3]
```diff
@@ -5,48 +5,104 @@ import Foundation
 
 class HttpResponseSpec: QuickSpec {
     override func spec() {
-        describe("HttpResponse") {
-            var responseBuilder = ResponseBuilder()
+        describe("A Reponse object") {
+            func summonBuilder() -> HttpResponseBuilder {
```
Comment (ericmeyer) [inline Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift]: The name `summonBuilder` is a bit odd. Although also weird, the name `buildBuilder` is both accurate and consistent with your other tests.

File: Tests/SwerverTests/RouterSpec/RouterSpec.swift [hunk 4]
```diff
@@ -6,122 +6,170 @@ import Foundation
...
+            func buildRouter() -> Router {
+                return Router()
             }
 
-            context("when receive an invalid request object") {
-                let anInvalidRequest: HttpRequest? = nil
-                let expectedResponse = router.process(request: anInvalidRequest)
-                
+            context("receives an invalid request object") {
                 it("returns a response with a 404 status code") {
```
Comment (ericmeyer) [inline Tests/SwerverTests/RouterSpec/RouterSpec.swift:14]: Try updating these tests to follow the Arrange/Act/Assert pattern. See here: http://wiki.c2.com/?ArrangeActAssert

File: Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift [hunk 5]
```diff
@@ -5,48 +5,104 @@ import Foundation
...
-                    .with(contentType: .text)
-                    .with(body: String())
-                    .build()
-                
-                expect(responseOne).to(equal(responseTwo))
+            context("when compares with another response object") {
+                it("is true when attributes match") {
+                    var responseBuilder = summonBuilder()
+                    
+                    let responseOne = TestData.buildResponse()
```
Comment (ericmeyer) [inline Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift]: I like the idea of abstracting details away from the test, however I think this takes it a bit too far. The problem using `buildResponse` here is that it hides the details that the test cares about.

In this case, you are saying that the attributes match, however, one would have to go to `TestData` to see that. I would actually create both responses directly in this test.

File: Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift [hunk 6]
```diff
@@ -5,48 +5,104 @@ import Foundation
...
+                        .with(body: "Hello")
+                        .build()
+                    
+                    expect(responseOne).to(equal(responseTwo))
+                }
+                
+                it("is false when the status code does not match") {
+                    var responseBuilder = summonBuilder()
+                    
+                    let responseOne = TestData.buildResponse()
```
Comment (ericmeyer) [inline Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift]: This test has two issues:
1. Details we care about are hidden from us
   * The differing status code
2. Details we do not directly care about are making the test noisy.
   * All the extra attributes

In other words, right now, you say the `"status code does not match"`, but it is hard to see that in the test.

## Examples

Here are two different examples of possible refactorings. Both of these changes make it so the test exposes the details you care about (the differing status code), and abstracts away or removes the details you do not.

### Removing Extraneous Details

This example removes the details we not care about. If some of the attributes are required, this does not work.

```swift
it("is false when the status code does not match") {
    var responseBuilder = summonBuilder()
    let responseOne = responseBuilder
      .with(statusCode: .not_found)
      .build()
    let responseTwo = responseBuilder
      .with(statusCode: .ok)
      .build()

    expect(responseOne).toNot(equal(responseTwo))
}
```

### Abstracting Extraneous Details

If you need or want to provide default values for some of the attributes, you can leverage the builder pattern here.

```swift
func defaultResponseBuilder() -> HttpResponseBuilder {
    return HttpResponseBuilder()
        .with(statusCode: .not_found)
        .with(contentType: .text)
        .set(header: .allow, value: "GET")
        .with(body: "Hello")
        .build()
}

it("is false when the status code does not match") {
    let responseOne = defaultResponseBuilder()
      .with(statusCode: .not_found)
      .build()
    let responseTwo = defaultResponseBuilder()
      .with(statusCode: .ok)
      .build()

    expect(responseOne).toNot(equal(responseTwo))
}
```

File: Tests/SwerverTests/TestData.swift [hunk 7]
```diff
@@ -8,4 +9,29 @@ class TestData {
...
+    
+    static func methodOptionsUrl() -> URL {
+        return URL(string: Resource.test_option.rawValue)!
+    }
+
+    static func notFoundUrl() -> URL {
+        return URL(string: "/not_here")!
+    }
+    
+    static func buildResponse() -> HttpResponse {
```
Comment (ericmeyer) [inline Tests/SwerverTests/TestData.swift]: I'm not sure I like this abstraction. For most of the tests I saw using it, it hid details that the test cared about.

**Review by ericmeyer — CHANGES_REQUESTED** [review]
Overall, very good changes. My main point of feedback is with `TestData.buildResponse`. In short, I'm not sure it makes sense to have that abstraction.

File: Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift (hunk 6 above)
Comment (ericmeyer) [reply Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift]: Oops! `build` should not be called in `defaultResponseBuilder`.

File: Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift (hunk 1 above)
Comment (mrfishball) [reply Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift:8]: 😄

File: Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift (hunk 2 above)
Comment (mrfishball) [reply Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift]: will take a look

File: Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift (hunk 3 above)
Comment (mrfishball) [reply Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift]: will change that

File: Tests/SwerverTests/RouterSpec/RouterSpec.swift (hunk 4 above)
Comment (mrfishball) [reply Tests/SwerverTests/RouterSpec/RouterSpec.swift:14]: will take a look

File: Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift (hunk 5 above)
Comment (mrfishball) [reply Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift]: will fix that

File: Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift (hunk 6 above)
Comment (mrfishball) [reply Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift]: will fix that

File: Tests/SwerverTests/TestData.swift (hunk 7 above)
Comment (mrfishball) [reply Tests/SwerverTests/TestData.swift]: will fix that

**Comment by ericmeyer** [comment]
@mrfishball Why are there two commits with the message "fix merge conflicts" ?

File: Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift (hunk 2 above)
Comment (mrfishball) [reply Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift]: fixed

File: Tests/SwerverTests/RouterSpec/RouterSpec.swift (hunk 4 above)
Comment (mrfishball) [reply Tests/SwerverTests/RouterSpec/RouterSpec.swift:14]: implemented

File: Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift [hunk 8]
```diff
@@ -5,48 +5,113 @@ import Foundation
...
+                    
+                    expect(responseOne).to(equal(responseTwo))
+                }
                 
-                let responseTwo = responseBuilder
-                    .with(statusCode: .ok)
-                    .with(contentType: .text)
-                    .with(body: String())
-                    .build()
+                it("is false when the status code does not match") {
```
Comment (ericmeyer) [inline Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift:30]: This test still has the problems mentioned in my previous comment regarding details in the test.
1. There is no way to know that the status code is different just from looking at the test.
2. There are extraneous details in the test that the test does not care about.

File: Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift [hunk 9]
```diff
@@ -5,48 +5,113 @@ import Foundation
...
-        describe("HttpResponse") {
-            var responseBuilder = ResponseBuilder()
+        describe("A Reponse object") {
+            func buildBuilder() -> HttpResponseBuilder {
+                return HttpResponseBuilder()
+            }
             
-            it("can equate two response object") {
-                let responseOne = responseBuilder
+            func defaultResponseBuilder() -> HttpResponse {
```
Comment (ericmeyer) [inline Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift]: This should be a return type of `HttpResponseBuilder`.

File: Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift [hunk 10]
```diff
@@ -5,48 +5,113 @@ import Foundation
...
+                return HttpResponseBuilder()
                     .with(statusCode: .ok)
                     .with(contentType: .text)
-                    .with(body: String())
+                    .set(header: .allow, value: "GET")
+                    .with(body: "Hello")
                     .build()
```
Comment (ericmeyer) [inline Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift]: This line should be removed. That way, existing tests can continue setting attributes.

**Review by ericmeyer — CHANGES_REQUESTED** [review]

File: Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift (hunk 8 above)
Comment (mrfishball) [reply Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift:30]: fixed

File: Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift (hunk 9 above)
Comment (mrfishball) [reply Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift]: fixed

File: Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift (hunk 10 above)
Comment (mrfishball) [reply Tests/SwerverTests/ResponseSpec/HttpResponseSpec.swift]: fixed

**Review by ericmeyer — APPROVED** [review]

**Comment by cmvandrevala** [comment]
Latest changes look good to me

**Pull request merged**